v2/tools/generator/internal/codegen/pipeline/assemble_oneof.go

```diff
@@ -0,0 +1,418 @@
+	commonProperties := oa.findCommonProperties(parent)
+	err := oa.updateOneOf(
+		leaf,
+		func(oneOf *astmodel.OneOfType) (*astmodel.OneOfType, error) {
+			result := oneOf
+			for _, obj := range commonProperties {
+				result = result.WithAdditionalPropertyObject(obj)
+			}
+
+			return result, nil
+		})
+
+	return errors.Wrapf(err, "embedding common properties from OneOf in leaf %s", leaf)
+}
+
+// findCommonProperties finds all the object properties referenced by name to embed in a leaf OneOf.
+// Recursively walking the reference tree is handled elsewhere.
+func (oa *oneOfAssembler) findCommonProperties(name astmodel.TypeName) []*astmodel.ObjectType {
+	if parentOneOf, ok := oa.asOneOf(name); ok {
+		// Parent is a OneOf
+		return parentOneOf.PropertyObjects()
+	}
+
+	if parentAllOf, ok := oa.asAllOf(name); ok {
+		// Parent is an AllOf
+		result := make([]*astmodel.ObjectType, 0, parentAllOf.Types().Len())
+		parentAllOf.Types().ForEach(func(t astmodel.Type, ix int) {
+			if obj, ok := astmodel.AsObjectType(t); ok {
+				result = append(result, obj)
+			}
+		})
+
+		return result
+	}
+
+	if obj, ok := oa.asObject(name); ok {
+		return []*astmodel.ObjectType{obj}
+	}
+
+	return nil
+}
+
+// addLeafReferenceToRoot adds a reference to the leaf to the root.
+// root is the name of the root of the oneOf hierarchy.
+// leaf is the name of the leaf to add a reference to.
+func (oa *oneOfAssembler) addLeafReferenceToRoot(root astmodel.TypeName, leaf astmodel.TypeName) error {
+	err := oa.updateOneOf(
+		root,
+		func(oneOf *astmodel.OneOfType) (*astmodel.OneOfType, error) {
+			return oneOf.WithType(leaf), nil
+		})
+
+	return errors.Wrapf(err, "adding leaf reference %s to root %s", leaf, root)
+}
+
+func (oa *oneOfAssembler) addDiscriminatorProperty(name astmodel.TypeName, rootName astmodel.TypeName) error {
+	// Find the name of the discriminator property from the root
+	root, ok := oa.asOneOf(rootName)
+	if !ok {
+		return errors.Errorf("couldn't find root %s", rootName)
+	}
+
+	discriminatorProperty := root.DiscriminatorProperty()
+	propertyName := oa.idFactory.CreatePropertyName(discriminatorProperty, astmodel.Exported)
+	propertyJson := oa.idFactory.CreateIdentifier(discriminatorProperty, astmodel.NotExported)
+
+	err := oa.updateOneOf(
+		name,
+		func(oneOf *astmodel.OneOfType) (*astmodel.OneOfType, error) {
+			// Create a name for the discriminator value
+			discriminatorValue := oneOf.DiscriminatorValue()
+			valueName := oa.idFactory.CreateIdentifier(discriminatorValue, astmodel.Exported)
+
+			// Create the discriminator property as a single valued enum
+			property := astmodel.NewPropertyDefinition(
+				propertyName,
+				propertyJson,
+				astmodel.NewEnumType(
+					astmodel.StringType,
+					astmodel.MakeEnumValue(valueName, fmt.Sprintf("%q", discriminatorValue))))
+
+			if discriminatorValue == "" {
+				klog.Warning("Weirdness")
```

Comment:
Is this a sufficient log? Maybe a `panic` if this doesn't actually happen now? Or else improve the logs warning so it's more clear.

Reply:
You're very polite. That's a debugging wart that no longer triggers. Will remove.